### PR TITLE
Clears the whole atlas at the same time

### PR DESCRIPTION
**Pull request by Kuranes**
Clears whole texture at once for shadowAtlas Avoid samsung bug on scissor+glClear
 Fixes shadowAtlas recent light and compiler refactors broke it.

### CODE REVIEW

File: sources/osg/RenderStage.js [hunk 1]
```diff
@@ -265,10 +265,21 @@ MACROUTILS.createPrototypeObject(
             state.applyAttribute(this.viewport);
 
             // fragment clipping
+
             if (this.camera) {
```
Comment (cedricpinson) [inline sources/osg/RenderStage.js]: All this the code in this.camera is useless, it was necessary if we cleared by sub regions but if we dont do that we dont need to apply it here

File: sources/osgShadow/ShadowMapAtlas.js [hunk 2]
```diff
@@ -314,10 +382,60 @@ MACROUTILS.createPrototypeObject(
...
+                if (this._shadowMaps.indexOf(shadowMap) !== -1) return;
+            }
+
+            this._shadowMaps.push(shadowMap);
+
+            var light = shadowMap.getLight();
+            if (this._lights.indexOf(light) === -1) {
+                this._lights.push(light);
+            } else {
+                console.assert(false);
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMapAtlas.js]: uncool

File: sources/osgShadow/ShadowMapAtlas.js [hunk 3]
```diff
@@ -314,10 +382,60 @@ MACROUTILS.createPrototypeObject(
...
             }
+
+            if (this._shadowedScene) shadowMap.setShadowedScene(this._shadowedScene);
+
+            var lightNumberArray = this._texture.getLightNumberArray();
+            if (lightNumberArray.indexOf(shadowMap.getLightNumber()) === -1) {
+                lightNumberArray.push(light.getLightNumber());
+            } else {
+                console.assert(false);
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowMapAtlas.js]: double uncool

File: sources/osgShadow/ShadowTextureAtlas.js [hunk 4]
```diff
@@ -20,10 +20,10 @@ var ShadowTextureAtlas = function() {
     this._uniforms = {};
     this._lightNumberArray = []; // default for a valid cloneType
 
-    this._viewMatrices = [];
-    this._projectionMatrices = [];
-    this._depthRanges = [];
-    this._mapSizes = [];
+    this._viewMatrices = {};
```
Comment (cedricpinson) [inline sources/osgShadow/ShadowTextureAtlas.js]: why switching [] to {}

File: sources/osgShadow/ShadowMapAtlas.js (hunk 3 above)
Comment (Kuranes) [reply sources/osgShadow/ShadowMapAtlas.js]: ha double sorry 
that debug code to double check error in lightnumber/index was helpful, but definitely not needed now

File: sources/osgShadow/ShadowTextureAtlas.js (hunk 4 above)
Comment (Kuranes) [reply sources/osgShadow/ShadowTextureAtlas.js]: As we're now indexing by lightnumbers now, not using lightArrayIndex, so it's a sparse array
if user put lights 8,9,10, better use an associative array ?

**Comment by Kuranes** [comment]
- Fixed all comments

File: sources/osgShadow/ShadowTextureAtlas.js (hunk 4 above)
Comment (cedricpinson) [reply sources/osgShadow/ShadowTextureAtlas.js]: no idea what is better in this case, but simpler is probably better

Comment (cedricpinson) [reply sources/osgShadow/ShadowTextureAtlas.js]: I would probably keep array because size will not be more up to 8 I guess

**Comment by Kuranes** [comment]
- rebased with latest
- use array for textureAtlas uniforms

**Pull request merged**